### ISSUE TITLE
[XLA:GPU] ScatterDeterminismExpander writes an out-of-bounds scatter index into the next row instead of dropping it

### ISSUE THREAD

**Issue by AshishKumar4** [openxla/xla#49380]
## Summary

With `--xla_gpu_deterministic_ops=true`, an out-of-bounds index in a multi-dimensional scatter can be written into the next row instead of being dropped. The index must not be a compile-time constant, and it must share a row-major position with another update in the same scatter. JAX lowers `mode="drop"` to a scatter that should skip every out-of-bounds index. On GPU without the flag, and on CPU, the result is correct.

## Reproducer

```python
# XLA_FLAGS=--xla_gpu_deterministic_ops=true JAX_PLATFORMS=cuda python repro.py
import jax
import jax.numpy as jnp

scatter = jax.jit(lambda buffer, rows, columns, values: buffer.at[rows, columns].set(values, mode="drop"))
print(scatter(jnp.zeros((2, 4), jnp.int32), jnp.arange(2)[:, None], jnp.array([[0, 4], [0, 4]]),
              jnp.array([[1, 2], [3, 4]], jnp.int32)).tolist())
```

| run | output |
|---|---|
| expected (and GPU without the flag, CPU) | `[[1, 0, 0, 0], [3, 0, 0, 0]]` |
| GPU, `--xla_gpu_deterministic_ops=true` | `[[1, 0, 0, 0], [2, 0, 0, 0]]` |

Row 0 writes `1` at column 0 and `2` at column 4, which is out of bounds and should be dropped. Row 1 writes `3` at column 0. With the flag, the value `2` lands at `(1, 0)` and overwrites row 1's own write.

Environment: stock jax 0.11.2 from PyPI (`pip install 'jax[cuda12]'`), jaxlib 0.11.2, jax-cuda12-plugin 0.11.2, NVIDIA L4 (Colab). The same wrong token appears in a larger program on RTX 3090 and T4.

## Analysis

The pattern matches `ScatterDeterminismExpander` that flattens the multi-dimensional scatter index row-major before sorting and combining updates, and only drops a flattened position that is outside the whole operand. The out-of-bounds index `(0, 4)` flattens to `0 * 4 + 4 = 4`, which is the position of `(1, 0)`. After the rewrite it is treated as an in-bounds duplicate of row 1's update. Each index needs to be bounds-checked per dimension before it is flattened.

These cases behave correctly:

- An out-of-bounds index whose flattened position lies outside the whole operand, for example column `8` (or `2**30`) in the example above.
- An out-of-bounds index whose flattened position collides with no update in the same scatter.
- Compile-time-constant indices, which are folded before the rewrite.

Related: #31746 changed the padded index value in the same pass's prefix scan; this is a different path, the per-dimension bounds of the original indices.

## Impact

`mode="drop"` with a sentinel of "one past the end of the axis" is the usual JAX way to mask entries out of a batched scatter: KV-cache writes of padded positions, per-row token buffers, logits masks. Under the deterministic flag those writes corrupt the next batch row. In our case a batched speculative decoder emitted row 0's token as row 1's first token, and a left-padded row's KV-cache write landed in the next row.

## Workaround

Send dropped entries to an index past the whole flattened index space, for example `2**30`, not to the axis size.
